fuel-core/src/tx_pool.rs

```diff
@@ -83,9 +83,44 @@ impl TxPool {
 
     pub async fn submit_tx(&self, tx: Transaction) -> Result<Bytes32, Error> {
         let db = self.db.clone();
-        let tx_id = tx.id();
+        // TxPool code flows and prepare tx for execution
+        let mut tx_to_exec = tx.clone();
+
+        if tx_to_exec.metadata().is_none() {
+            tx_to_exec.precompute_metadata();
+        }
+
+        self.fuel_txpool
+            .insert(vec![Arc::new(tx_to_exec.clone())])
+            .await;
```

Comment:
You need to check the result of the `insert`, tx needs to be valid to be included inside the pool.
Maybe we can make the `insert` and `insert_vec` functions to better facilitate this.

Reply:
So on the tests that currently require the fallback (ie don't work with the new txpool) it is because I am running into the `[Err(Transaction is not inserted. UTXO is not existing: 0x99dd7fc1ad584d9b174275ef9de7bda04fc61e38899fdce22fd31a49f3fc47d6)]` error, which seems to stem from `TestContext`'s `transfer` method generating a `utxo_id:` via `self.rng.gen()`, so when `precompute_metadata` is called it requests an input transaction which doesn't exist causing it to not be inserted.

Or do you just mean runtime validation that inserting was successful?

Reply:
yea, there are not check there and data can't be random, it is expected for them to fail. Changing those test to reflect error seems resonable.

runtime validation is needed here, you need to return error there for user to know what is happening.

Reply:
Hmm, we've been spoofing utxos in tests in the early days, but it seems like the txpool actually validates their existence. Currently in other parts of the system, we validate utxo existence only when the `utxo-validation` flag is enabled. This is because full validation will be a fairly breaking change to all the downstream test suites in the SDK and in sway, so we've been holding off on enabling `utxo-validation` by default until they are adapted to initialize coins properly. 

Given that integrating this new txpool would conflict with our current `utxo-validation` default being `false`, we should gate usage of the new txpool using the same `utxo-validation` feature flag. That way we can begin writing tests with utxo-validation enabled, without breaking everything downstream.

Reply:
Another note is that the utxo lookup mechanism used by the txpool will need to be adjusted, since it looks up coins only from other pre-existing transactions, instead of directly checking the utxo set. This will cause issues for us since we need to be able to add coins directly to the utxo set from the genesis block without an associated transaction.

Reply:
Would moving this change to a separate feature mean making a separate function like `submit_tx_with_validation` and then only making that function available under a feature, or is there a way to overload a function based on feature?

Will work on adding runtime validation in the meantime though

Reply:
Checking the feature flag should be as simple as something like this:

```
        if self.executor.config.utxo_validation {
            // your new txpool integration logic   
        } else {
            // the legacy dummy logic from before
        }

```

Reply:
I'll make a new pr to fix the utxo lookup issue I described above

Reply:
@Voxelot Yea, i should have used utxo db table for this, it seems like straightforward fix, if you want I can do it.

@ControlCplusControlV in general we should probably change error type in fuel-core to return `anyhow::Error` as we dont care what error is we just want to propagate it to user. But for `fuel-txpool` i feel like we need to chainge it to proper error type. 
In rust you have two main error libs `thiserror` and `anyhow` this is good read: https://github.com/dtolnay/anyhow

There is good comparison there:
```
Use Anyhow if you don't care what error type your functions return, you just want it to be easy. This is common in
application code. Use [thiserror](https://github.com/dtolnay/thiserror) if you are a library that wants to design your own 
dedicated error type(s) so that on failures the caller gets exactly the information that you choose.
```

Reply:
In the context of the graphql function, we should be able to just use the `?` operator to make the error types match as long as the error has a display trait enabled.

Reply:
@rakita feel free to go ahead and make the fix, I haven't started on it yet.

As a side-note, is there any need for UtxoId to be still be txid + out_idx after this fix? It may help consolidate things like deposit coins vs utxos into the same db table if the id was just a simple 32 byte hash.

Reply:
will make bug issue.

txid+out_idx is still relevent and helpfull for dependencies